From 14ad271ebc76467de7a6fabe6401c40596d4198e Mon Sep 17 00:00:00 2001
Message-Id: <14ad271ebc76467de7a6fabe6401c40596d4198e.1413836944.git.chang-joon.lee@intel.com>
In-Reply-To: <5d8aa9994fff43965a6fab8c9a528b8b47a9d624.1413836944.git.chang-joon.lee@intel.com>
References: <5d8aa9994fff43965a6fab8c9a528b8b47a9d624.1413836944.git.chang-joon.lee@intel.com>
From: Deepak S <deepak.s@intel.com>
Date: Wed, 15 Oct 2014 18:47:39 +0530
Subject: [PATCH 68/71] FOR_UPSTREAM [VPG]: drm/i915: Take rps lock before
 accessing vlv punit read/write.

During maxfifo programming, We missed taking rps mutex lock before
accessing vlv punit read/write, which introduced a WARN_ON Stack dump.

Issue: GMINL-2375
Change-Id: I8c54e678f92fdbe2909813bdd424f251a9086170
Signed-off-by: Deepak S <deepak.s@intel.com>
---
 drivers/gpu/drm/i915/intel_pm.c |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index cc6dd50..bb15dbe 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -1639,18 +1639,22 @@ void intel_update_maxfifo(struct drm_i915_private *dev_priv)
 			&& !dev_priv->maxfifo_enabled) {
 			I915_WRITE(FW_BLC_SELF_VLV, FW_CSPWRDWNEN);
 			if (IS_CHERRYVIEW(dev_priv->dev)) {
+				mutex_lock(&dev_priv->rps.hw_lock);
 				val = vlv_punit_read(dev_priv, CHV_DPASSC);
 				vlv_punit_write(dev_priv, CHV_DPASSC,
 						(val | CHV_PW_MAXFIFO_MASK));
+				mutex_unlock(&dev_priv->rps.hw_lock);
 			}
 			dev_priv->maxfifo_enabled = true;
 		} else if (dev_priv->maxfifo_enabled &&
 				!single_plane_enabled(dev_priv->plane_stat)) {
 			I915_WRITE(FW_BLC_SELF_VLV, ~FW_CSPWRDWNEN);
 			if (IS_CHERRYVIEW(dev_priv->dev)) {
+				mutex_lock(&dev_priv->rps.hw_lock);
 				val = vlv_punit_read(dev_priv, CHV_DPASSC);
 				vlv_punit_write(dev_priv, CHV_DPASSC,
 					(val & ~(CHV_PW_MAXFIFO_MASK)));
+				mutex_unlock(&dev_priv->rps.hw_lock);
 			}
 			dev_priv->maxfifo_enabled = false;
 		}
-- 
1.7.9.5

